figure2.py

```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/python
+
+from subprocess import call
+from collections import namedtuple
+
+UPLINK_EXT = 'Verizon-LTE-short.up'
+UPLINK = '~/ABC-1/mahimahi/traces/{}'.format(UPLINK_EXT)
+DOWNLINK = '~/ABC-1/bw48.mahi'
+DELAY = 50
+
+Protocol = namedtuple('Protocol', ['name', 'pre_commands', 'post_commands',
+    'uplink_queue', 'uplink_queue_args', 'commands'])
+Stats = namedtuple('Stats', ['util', 'delay'])
+
+ABC = Protocol(
+    name='abc',
+    pre_commands=['python server.py &'],
+    post_commands=['pkill -f server.py'],
+    uplink_queue='cellular',
+    # TODO: make packets consistent between protocols...
+    uplink_queue_args='packets=250,qdelay_ref=50,beta=75',
+    commands='python client.py'
+)
+CUBIC = Protocol(
+    name='cubic',
+    pre_commands=['iperf -s -p 42425 -w 16m &'],
+    post_commands=['killall iperf'],
+    uplink_queue='droptail',
+    uplink_queue_args='packets=100',
+    commands='./start-tcp.sh cubic'
+)
+# TODO: fix vegas!
+# Attempt to set 'vegas' congestion control failed: No such file or directory
+VEGAS = CUBIC._replace(
+    name='vegas',
+    commands='./start-tcp.sh vegas'
+)
+CUBIC_CODEL = CUBIC._replace(
+    name='cubiccodel',
+    uplink_queue='codel',
+    uplink_queue_args='packets=100,target=50,interval=100'
+)
+CUBIC_PIE = CUBIC._replace(
+    name='cubicpie',
+    uplink_queue='pie',
+    uplink_queue_args='packets=100,qdelay_ref=50,max_burst=100'
+)
+# TODO: fix BBR!
+# qdisc fq 8001: root refcnt 2 limit 10000p flow_limit 100p buckets 1024 quantum 3000 initial_quantum 15000
+# Attempt to set 'bbr' congestion control failed: No such file or directory
+BBR = CUBIC._replace(
+    name='bbr',
+    commands='./start-tcp.sh bbr'
+)
+
+PROTOS = [ABC, CUBIC, CUBIC_CODEL, CUBIC_PIE]
+stats = dict()
+
+def run_exp(proto, skip=False):
+    uplink_log_file = 'logs/{}-{}.log'.format(UPLINK_EXT, proto.name)
+    results_file = 'results/{}-{}.txt'.format(UPLINK_EXT, proto.name)
+
+    if not skip:
+        print('Running {} experiment...'.format(proto.name))
+
+        exp = 'mm-delay {delay} mm-link --uplink-log={uplink_log} \
+            --uplink-queue={uplink_queue} --uplink-queue-args=\"{uplink_queue_args}\" \
+            {uplink} {downlink} {commands}'.format(delay=DELAY, uplink_log=uplink_log_file,
+                uplink_queue=proto.uplink_queue, uplink_queue_args=proto.uplink_queue_args,
+                uplink=UPLINK, downlink=DOWNLINK, commands=proto.commands)
+        results = 'mm-throughput-graph {delay} {log_file} > /dev/null 2> {results_file}'.format(
+            delay=DELAY, log_file=uplink_log_file, results_file=results_file)
+
+        commands = list(proto.pre_commands)
+        commands.append(exp)
+        commands.extend(proto.post_commands)
+        commands.append(results)
+        for c in commands:
+            call(c, shell=True)
+
+    with open(results_file) as f:
+        lines = f.readlines()
```

Comment:
We could probably save as a json for nicer parsing?  Don't know if that's an option with the current pipeline

Reply:
You mean save the figure2 results as json? I liked csv because we can easily import it into a spreadsheet to create a graph that way, but if you feel strongly about it we can save as json. This part is parsing the `mm-throughput-graph`.

Reply:
I don't feel strongly about it.  If csv is more convenient that's also fine